### PR TITLE
Operational Learning Summary Generation

### PR DESCRIPTION
**Pull request by susilnem**
Addresses
- https://github.com/IFRCGo/go-api/issues/2198
- https://github.com/IFRCGo/go-api/issues/2199
- https://github.com/IFRCGo/go-api/issues/2200
- https://github.com/IFRCGo/go-api/issues/2201

## Changes
- Need to install poetry
- Run command `python manage.py update_translation_fields` for updating the translation.

## Checklist
Things that should succeed before merging.

- [ ] Updated/ran unit tests
- [ ] Updated CHANGELOG.md

## Release

If there is a version update, make sure to tag the repository with the latest version.

### CODE REVIEW

File: per/drf_views.py [hunk 1]
```diff
@@ -701,6 +717,21 @@ class Meta:
             "appeal_code__region": ("exact", "in"),
         }
 
+    def get_cache_response(self, queryset, name, value):
+        if ops_learning_cache_response := OpsLearningCacheResponse.objects.filter(id=value).first():
```
Comment (thenav56) [inline per/drf_views.py:721]: Let's also add additional check
```suggestion
        if value and (ops_learning_cache_response := OpsLearningCacheResponse.objects.filter(id=value).first()):
```

File: per/serializers.py [hunk 2]
```diff
@@ -1046,6 +1049,29 @@ class Meta:
         exclude = ("learning", "type", "organization", "sector", "per_component")
 
 
+class AppealDetailSerializer(serializers.ModelSerializer):
+    atype = serializers.SerializerMethodField()
```
Comment (thenav56) [inline per/serializers.py:1053]: Maybe use this instead? SerializerMethodField doesn't provide proper schema
```suggestion
    atype_display = serializers.CharField(source='get_atype_display', read_only=True)
```

File: per/serializers.py [hunk 3]
```diff
@@ -1118,3 +1144,60 @@ class Meta:
...
+        fields = [
+            "id",
+            "content",
+            "title",
+            "extract_count",
+            "sources",
+        ]
+
+    def get_extract_count(self, obj) -> int:
+        return obj.used_ops_learning.count()
```
Comment (thenav56) [inline per/serializers.py:1165]: N+1?

File: per/ops_learning_summary.py [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+        "1. Describe, Summarize and Compare: Identify and detail the who, what, where and when"
+        "2. Explain and Connect: Analyze why events happened and how they are related"
+        "3. Identify gaps: Assess what data is available, what is missing and potential biases"
+        "4. Identify key messages: Determine if there are important stories and signals hidden in the data"
+        "5. Conclude and make your case"
+    )
+
+    client = AzureOpenAI(
+        azure_endpoint=settings.AZURE_OPENAI_ENDPOINT, api_key=settings.AZURE_OPENAI_KEY, api_version="2023-05-15"
+    )
```
Comment (thenav56) [inline per/ops_learning_summary.py:122]: Use cached_property for this

Add a `TODO: to create a separate object for integration with Open AI for common use cases`

File: per/task.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        )
+        OpsLearningSummaryTask.get_or_create_summary(
+            ops_learning_summary_instance, primary_learning_prompt, secondary_learning_prompt
+        )
+    except Exception as e:
+        OpsLearningSummaryTask.change_ops_learning_status(
+            instance=ops_learning_summary_instance, status=OpsLearningCacheResponse.Status.FAILED
+        )
+        logger.error(e)
+        raise e
```
Comment (thenav56) [inline per/task.py:31]:
```suggestion
        # TODO: Add extra context for which ops learning cache response failed
        logger.error("Ops learning process failed", exc_info=True)
        return False
   return True
```

File: per/task.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+from api.logger import logger
+from per.models import OpsLearningCacheResponse
+from per.ops_learning_summary import OpsLearningSummaryTask
+
+
+@shared_task
+def generate_summary(ops_learning_summary_id: int, filter_data: dict):
+    try:
+        ops_learning_summary_instance = OpsLearningCacheResponse.objects.filter(id=ops_learning_summary_id).first()
```
Comment (thenav56) [inline per/task.py:11]: move this outside try
```suggestion
        ops_learning_summary_instance = OpsLearningCacheResponse.objects.filter(id=ops_learning_summary_id).first()
        if ops_learning_summary_instance is None:
            # TODO: Add extra context for which ops learning cache response failed
            logger.error("Missing Ops learning cache response triggered")
            return False
```

**Review by thenav56 — CHANGES_REQUESTED** [review]
Round 1

File: per/serializers.py (hunk 2 above)
Comment (susilnem) [reply per/serializers.py]: This Serializer is not in use.

**Review by thenav56 — APPROVED** [review]

**Comment by thenav56** [comment]
@szabozoltan69 When you have time, please review and merge this to staging. Thanks

**Review by szabozoltan69 — APPROVED** [review]
Nice one, the performance ones (prefetch_related...) are great code parts!

**Comment by susilnem** [comment]
@szabozoltan69 
We also need to execute the following command:  
`python manage.py update_translation_fields`

**Comment by szabozoltan69** [comment]
Done (run for 3 minutes). Sorry for not noting it at first. @susilnem

**Pull request merged**